source/documentBase.py

```diff
@@ -277,106 +322,152 @@ def _getFirstOrLastTableCell(
-		# to allow users being able to skip merged cells without affecting the initial column/row index.
-		# For more info see issue #11919 and #7278.
-		if (
-			self._lastTableSelection
-			and origRow == self._lastTableSelection.lastRow
-			and origCol == self._lastTableSelection.lastCol
-			and self._lastTableSelection.axis == axis
-		):
-			if axis == _Axis.ROW:
-				origCol = self._lastTableSelection.trueCol
-				origColSpan = self._lastTableSelection.colSpan
-			else:
-				origRow = self._lastTableSelection.trueRow
-				origRowSpan = self._lastTableSelection.rowSpan
+			raise e
 
 		try:
 			if movement in {_Movement.PREVIOUS, _Movement.NEXT}:
 				info = self._getNearestTableCell(
-					tableID,
 					self.selection,
-					origRow,
-					origCol,
-					origRowSpan,
-					origColSpan,
+					cell,
 					movement,
-					axis
+					axis,
 				)
 			elif movement in {_Movement.FIRST, _Movement.LAST}:
 				info = self._getFirstOrLastTableCell(
-					tableID,
 					self.selection,
-					origRow,
-					origCol,
-					origRowSpan,
-					origColSpan,
+					cell,
 					movement,
 					axis
 				)
+			elif movement is None:
+				info = self._getTableCellAt(cell.tableID, self.selection, cell.row, cell.col)
 			else:
 				raise ValueError(f"Unknown movement {movement}")
-			newTableID, newRow, newCol, newRowSpan, newColSpan = self._getTableCellCoords(info)
-		except LookupError:
+			newCell = self._getTableCellCoords(info)
+		except LookupError as e:
+			if raiseOnEdge:
+				raise e
 			# Translators: The message reported when a user attempts to use a table movement command
 			# but the cursor can't be moved in that direction because it is at the edge of the table.
 			ui.message(_("Edge of table"))
 			# Retrieve the cell on which we started.
-			info = self._getTableCellAt(tableID, self.selection,origRow, origCol)
-			newTableID, newRow, newCol, newRowSpan, newColSpan = self._getTableCellCoords(info)
+			try:
+				info = self._getTableCellAt(cell.tableID, self.selection, cell.row, cell.col)
+			except LookupError as e:
+				raise RuntimeError("Unable to find current cell.", e)
+			newCell = self._getTableCellCoords(info)
+		tableSelection = _TableSelection(
+			lastRow=newCell.row,
+			lastCol=newCell.col,
+			axis=axis,
+			trueRow=cell.row if axis == _Axis.COLUMN else newCell.row,
+			rowSpan=cell.rowSpan if axis == _Axis.COLUMN else newCell.rowSpan,
+			trueCol=cell.col if axis == _Axis.ROW else newCell.col,
+			colSpan=cell.colSpan if axis == _Axis.ROW else newCell.colSpan,
+		) if movement is not None else None
+		return newCell, info, tableSelection
+
+	def _tableMovementScriptHelper(
+			self,
+			movement: _Movement = _Movement.NEXT,
+			axis: Optional[_Axis] = None
+	):
+		# documentBase is a core module and should not depend on these UI modules and so they are imported
+		# at run-time. (#12404)
+		from scriptHandler import isScriptWaiting
+		from speech import speakTextInfo
+
+		if isScriptWaiting():
+			return

```

Comment:
Suggested an extra new line here

```suggestion
		if isScriptWaiting():
			return

```